tests/smoke.rs

```diff
@@ -73,6 +71,49 @@ fn prop_max_streams() {
     quickcheck(prop as fn(_) -> _);
 }
 
+#[test]
+fn prop_send_recv_half_closed() {
+    fn prop(msg: u8) -> bool {
+        let (l, a) = bind();
+        let cfg = Config::default();
+
+        // Server should be able to write on a stream shutdown by the client.
+        let server = server(cfg.clone(), l).and_then(|c| c
+            .into_future()
+            .map_err(|(e,_)| e)
+            .and_then(|(stream, _)| {
+                let mut s = stream.expect("S: No incoming stream");
+                future::poll_fn(move || {
+                    let mut buf = vec![0;1];
+                    try_ready!(s.poll_read(&mut buf));
+                    assert!(s.state() == Some(State::RecvClosed));
+                    let w = s.write(&buf);
```

Comment:
This and the next line are synchronous calls and may fail with `std::io::ErrorKind::WouldBlock`.

Reply:
I am aware, but only if the stream has no credit, which I considered here as a genuine test failure and hence thought it is fine to write synchronously as it simplifies the test. Do you disagree, i.e. would want the test to write asynchronously?

Reply:
> I am aware, but only if the stream has no credit, which I considered here as a genuine test failure and hence thought it is fine to write synchronously as it simplifies the test.

If [flush](https://github.com/paritytech/yamux/blob/master/src/connection.rs#L716)  returns `WouldBlock` it has other reasons.

> Do you disagree, i.e. would want the test to write asynchronously?

Yes please. The test should only assume that the stream is non-blocking. The fact that the current implementation uses internal buffering and therefore only exhibits `WouldBlock` errors in a limited set of cases can change any time and is not something to rely on.

Reply:
I changed the test to be based on futures via the `tokio::io` functions instead of using `poll_fn`. Apart from no longer using blocking IO, the test should look cleaner now and uses not just single byte messages but test `Msg`s of random length.